The first error is:
```
Error while processing /__w/spectre/spectre/tests/Unit/DataStructures/Tensor/Expressions/Test_AddSubtract.cpp.
/__w/spectre/spectre/src/DataStructures/Tensor/Expressions/AddSubtract.hpp:14:10: error: 'DataStructures/Tensor/Expressions/NumberAsExpression.hpp' file not found
```
It couldn't find a file (maybe you forgot to commit it?). This means the rest of the errors are not useful since it can't really do an analysis of code that can't compile :)

--- a/src/DataStructures/Tensor/Expressions/AddSubtract.hpp
+++ b/src/DataStructures/Tensor/Expressions/AddSubtract.hpp
@@ -10,6 +10,8 @@
 #include <array>
 #include <cstddef>
 
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/NumberAsExpression.hpp"
 #include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
 #include "Utilities/Requires.hpp"
 #include "Utilities/TMPL.hpp"
@@ -57,11 +59,16 @@ template <typename T1, typename T2, template <typename...> class ArgsList1,
 struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
     : public TensorExpression<
           AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>,
-          typename T1::type,
+          std::conditional_t<
+              std::is_same<typename T1::type, DataVector>::value or
+                  std::is_same<typename T2::type, DataVector>::value,
+              DataVector, double>,
           tmpl::transform<typename T1::symmetry, typename T2::symmetry,
                           tmpl::append<tmpl::max<tmpl::_1, tmpl::_2>>>,
           typename T1::index_list, typename T1::args_list> {
-  static_assert(std::is_same<typename T1::type, typename T2::type>::value,
+  static_assert(std::is_same<typename T1::type, typename T2::type>::value or
+                    std::is_same<T1, NumberAsExpression>::value or
+                    std::is_same<T2, NumberAsExpression>::value,
                 "Cannot add or subtract Tensors holding different data types.");
   static_assert(
       detail::AddSubIndexCheck<typename T1::index_list, typename T2::index_list,
@@ -74,7 +81,10 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
                 "Invalid Sign provided for addition or subtraction of Tensor "
                 "elements. Sign must be 1 (addition) or -1 (subtraction).");
 
-  using type = typename T1::type;
+  using type =
+      std::conditional_t<std::is_same<typename T1::type, DataVector>::value or
+                             std::is_same<typename T2::type, DataVector>::value,
+                         DataVector, double>;
   using symmetry = tmpl::transform<typename T1::symmetry, typename T2::symmetry,
                                    tmpl::append<tmpl::max<tmpl::_1, tmpl::_2>>>;
   using index_list = typename T1::index_list;
@@ -113,12 +123,12 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
 /*!
  * \ingroup TensorExpressionsGroup
  */
-template <typename T1, typename T2, typename X, typename Symm1, typename Symm2,
-          typename IndexList1, typename IndexList2, typename Args1,
-          typename Args2>
+template <typename T1, typename T2, typename X1, typename X2, typename Symm1,
+          typename Symm2, typename IndexList1, typename IndexList2,
+          typename Args1, typename Args2>
 SPECTRE_ALWAYS_INLINE auto operator+(
-    const TensorExpression<T1, X, Symm1, IndexList1, Args1>& t1,
-    const TensorExpression<T2, X, Symm2, IndexList2, Args2>& t2) {
+    const TensorExpression<T1, X1, Symm1, IndexList1, Args1>& t1,
+    const TensorExpression<T2, X2, Symm2, IndexList2, Args2>& t2) {
   static_assert(tmpl::size<Args1>::value == tmpl::size<Args2>::value,
                 "Tensor addition is only possible with the same rank tensors");
   static_assert(tmpl::equal_members<Args1, Args2>::value,
@@ -127,15 +137,51 @@ SPECTRE_ALWAYS_INLINE auto operator+(
   return TensorExpressions::AddSub<T1, T2, Args1, Args2, 1>(~t1, ~t2);
 }
 
+// @{
+/// \ingroup TensorExpressionsGroup
+/// \brief Returns the tensor expression representing the sum of a tensor
+/// expression and a `double`
+///
+/// \details
+/// The tensor expression operand must represent an expression that, when
+/// evaluated, would be a rank 0 tensor. For example, if `R` and `S` are
+/// Tensors, here is a non-exhaustive list of some of the acceptable forms that
+/// the tensor expression operand could take:
+/// - `R()`
+/// - `R(ti_A, ti_a)`
+/// - `(R(ti_A, ti_B) * S(ti_a, ti_b))`
+///
+/// \tparam T the derived TensorExpression type of the tensor expression operand
+/// of the sum
+/// \tparam X the type of data stored in the tensor expression operand of the
+/// sum
+/// \param t the tensor expression operand of the sum
+/// \param number the `double` operand of the sum
+/// \return the tensor expression representing the sum of a tensor expression
+/// and a `double`
+template <typename T, typename X>
+SPECTRE_ALWAYS_INLINE auto operator+(
+    const TensorExpression<T, X, tmpl::list<>, tmpl::list<>, tmpl::list<>>& t,
+    const double number) {
+  return t + TensorExpressions::NumberAsExpression(number);
+}
+template <typename T, typename X>
+SPECTRE_ALWAYS_INLINE auto operator+(
+    const double number,
+    const TensorExpression<T, X, tmpl::list<>, tmpl::list<>, tmpl::list<>>& t) {
+  return TensorExpressions::NumberAsExpression(number) + t;
+}
+// @}
+
 /*!
  * \ingroup TensorExpressionsGroup
  */
-template <typename T1, typename T2, typename X, typename Symm1, typename Symm2,
-          typename IndexList1, typename IndexList2, typename Args1,
-          typename Args2>
+template <typename T1, typename T2, typename X1, typename X2, typename Symm1,
+          typename Symm2, typename IndexList1, typename IndexList2,
+          typename Args1, typename Args2>
 SPECTRE_ALWAYS_INLINE auto operator-(
-    const TensorExpression<T1, X, Symm1, IndexList1, Args1>& t1,
-    const TensorExpression<T2, X, Symm2, IndexList2, Args2>& t2) {
+    const TensorExpression<T1, X1, Symm1, IndexList1, Args1>& t1,
+    const TensorExpression<T2, X2, Symm2, IndexList2, Args2>& t2) {
   static_assert(tmpl::size<Args1>::value == tmpl::size<Args2>::value,
                 "Tensor addition is only possible with the same rank tensors");
   static_assert(tmpl::equal_members<Args1, Args2>::value,
@@ -143,3 +189,39 @@ SPECTRE_ALWAYS_INLINE auto operator-(
                 "occurs from expressions like A(_a, _b) - B(_c, _a)");
   return TensorExpressions::AddSub<T1, T2, Args1, Args2, -1>(~t1, ~t2);
 }
+
+// @{
+/// \ingroup TensorExpressionsGroup
+/// \brief Returns the tensor expression representing the difference of a tensor
+/// expression and a `double`
+///
+/// \details
+/// The tensor expression operand must represent an expression that, when
+/// evaluated, would be a rank 0 tensor. For example, if `R` and `S` are
+/// Tensors, here is a non-exhaustive list of some of the acceptable forms that
+/// the tensor expression operand could take:
+/// - `R()`
+/// - `R(ti_A, ti_a)`
+/// - `(R(ti_A, ti_B) * S(ti_a, ti_b))`
+///
+/// \tparam T the derived TensorExpression type of the tensor expression operand
+/// of the difference
+/// \tparam X the type of data stored in the tensor expression operand of the
+/// difference
+/// \param t the tensor expression operand of the difference
+/// \param number the `double` operand of the difference
+/// \return the tensor expression representing the difference of a tensor
+/// expression and a `double`
+template <typename T, typename X>
+SPECTRE_ALWAYS_INLINE auto operator-(
+    const TensorExpression<T, X, tmpl::list<>, tmpl::list<>, tmpl::list<>>& t,
+    const double number) {
+  return t - TensorExpressions::NumberAsExpression(number);
+}
+template <typename T, typename X>
+SPECTRE_ALWAYS_INLINE auto operator-(
+    const double number,
+    const TensorExpression<T, X, tmpl::list<>, tmpl::list<>, tmpl::list<>>& t) {
+  return TensorExpressions::NumberAsExpression(number) - t;
+}
+// @}

--- a/tests/Unit/DataStructures/Tensor/Expressions/Test_AddSubtract.cpp
+++ b/tests/Unit/DataStructures/Tensor/Expressions/Test_AddSubtract.cpp
@@ -7,13 +7,103 @@
 #include <iterator>
 #include <numeric>
 
+#include "DataStructures/DataVector.hpp"
 #include "DataStructures/Tensor/Expressions/AddSubtract.hpp"
 #include "DataStructures/Tensor/Expressions/Evaluate.hpp"
 #include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
 #include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace {
+template <typename... Ts>
+void assign_unique_values_to_tensor(
+    const gsl::not_null<Tensor<double, Ts...>*> tensor) noexcept {
+  std::iota(tensor->begin(), tensor->end(), 0.0);
+}
+
+template <typename... Ts>
+void assign_unique_values_to_tensor(
+    const gsl::not_null<Tensor<DataVector, Ts...>*> tensor) noexcept {
+  double value = 0.0;
+  for (auto index_it = tensor->begin(); index_it != tensor->end(); index_it++) {
+    for (auto vector_it = index_it->begin(); vector_it != index_it->end();
+         vector_it++) {
+      *vector_it = value;
+      value += 1.0;
+    }
+  }
+}
+
+// \brief Test the sum and difference of a `double` and tensor expression is
+// correctly evaluated
+//
+// \details
+// The cases tested are:
+// - \f$L = R + S\f$
+// - \f$L = R - S\f$
+// - \f$L = G^{i}{}_{i} + R\f$
+// - \f$L = G^{i}{}_{i} - R\f$
+// - \f$L = R + S + T\f$
+// - \f$L = R - G^{i}{}_{i} + T\f$
+//
+// where \f$R\f$ and \f$T\f$ are `double`s and \f$S\f$, \f$G\f$, and \f$L\f$
+// are Tensors with data type `double` or DataVector.
+//
+// \tparam DataType the type of data being stored in the tensor expression
+// operand of the sums and differences
+template <typename DataType>
+void test_addsub_double(const DataType& used_for_size) noexcept {
+  Tensor<DataType> S{{{used_for_size}}};
+  if (std::is_same_v<DataType, double>) {
+    // Replace tensor's value from `used_for_size` to a proper test value
+    S.get() = 2.4;
+  } else {
+    assign_unique_values_to_tensor(make_not_null(&S));
+  }
+
+  Tensor<DataType, Symmetry<2, 1>,
+         index_list<SpatialIndex<3, UpLo::Up, Frame::Grid>,
+                    SpatialIndex<3, UpLo::Lo, Frame::Grid>>>
+      G(used_for_size);
+  assign_unique_values_to_tensor(make_not_null(&G));
+
+  DataType G_trace = make_with_value<DataType>(used_for_size, 0.0);
+  for (size_t i = 0; i < 3; i++) {
+    G_trace += G.get(i, i);
+  }
+
+  // \f$L = R + S\f$
+  const Tensor<DataType> R_plus_S = TensorExpressions::evaluate(5.6 + S());
+  // \f$L = R - S\f$
+  const Tensor<DataType> R_minus_S = TensorExpressions::evaluate(1.1 - S());
+  // \f$L = G^{i}{}_{i} + R\f$
+  const Tensor<DataType> G_plus_R =
+      TensorExpressions::evaluate(G(ti_I, ti_i) + 8.2);
+  // \f$L = G^{i}{}_{i} - R\f$
+  const Tensor<DataType> G_minus_R =
+      TensorExpressions::evaluate(G(ti_I, ti_i) - 3.5);
+  // \f$L = R + S + T\f$
+  const Tensor<DataType> R_plus_S_plus_T =
+      TensorExpressions::evaluate(0.7 + S() + 9.8);
+  // \f$L = R - G^{i}{}_{i} + T\f$
+  const Tensor<DataType> R_minus_G_plus_T =
+      TensorExpressions::evaluate(5.9 - G(ti_I, ti_i) + 4.7);
+
+  CHECK(R_plus_S.get() == 5.6 + S.get());
+  CHECK(R_minus_S.get() == 1.1 - S.get());
+  CHECK(G_plus_R.get() == G_trace + 8.2);
+  CHECK(G_minus_R.get() == G_trace - 3.5);
+  CHECK(R_plus_S_plus_T.get() == 0.7 + S.get() + 9.8);
+  CHECK(R_minus_G_plus_T.get() == 5.9 - G_trace + 4.7);
+}
+}  // namespace
 
 SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.Expression.AddSubtract",
                   "[DataStructures][Unit]") {
+  test_addsub_double(std::numeric_limits<double>::signaling_NaN());
+  test_addsub_double(
+      DataVector(5, std::numeric_limits<double>::signaling_NaN()));
+
   // Test adding scalars
   const Tensor<double> scalar_1{{{2.1}}};
   const Tensor<double> scalar_2{{{-0.8}}};

